### PR TITLE
[FLINK-1723] [ml] [WIP] Add cross validation for model evaluation

### PR DESCRIPTION
**Pull request by thvasilo**
Cross validation (CV) [1] is a standard tool to estimate the test error for a model. As such it is a crucial tool for every machine learning library.

This builds upon the ongoing work on the evaluation framework for FlinkML.
As such, the current version supports calculating the score of Predictors only, however the end goal is to be able to have CV for Estimators as well to cover the unsupervised learning case.

We are using some code from the Apache Spark project, mostly simple routines for probabilistic sampling of datasets and generation of KFold CV data.

More and better tests need to be added to the implementation, and the current sampling approaches probably will not work if used within an iteration.

### CODE REVIEW

File: flink-staging/flink-ml/src/test/scala/org/apache/flink/ml/evaluation/CrossValidationITSuite.scala [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    import RegressionData._
+
+
+    val inputDS = env.fromCollection(data)
+
+    val mlr = MultipleLinearRegression()
+      .setStepsize(10.0)
+      .setIterations(100)
+
+    println()
```
Comment (tillrohrmann) [inline flink-staging/flink-ml/src/test/scala/org/apache/flink/ml/evaluation/CrossValidationITSuite.scala]: What does `println` do?

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        if (scorerOption.isEmpty) {
+          predictor.score(testing)
+        } else {
+          val s = scorerOption.get
+          s.evaluate(testing, predictor)
+        }
+    }
+    // TODO: Undecided on the return type: Array[DS[Double]] or DS[Double] i.e. reduce->union?
+    // Or: Return mean and std?
+    scores//.reduce((right: DataSet[Double], left: DataSet[Double]) => left.union(right)).mean()
```
Comment (tillrohrmann) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: I think that the mean would be a good return value. How do the other frameworks do it?

File: flink-staging/flink-ml/src/test/scala/org/apache/flink/ml/evaluation/CrossValidationITSuite.scala (hunk 1 above)
Comment (thvasilo) [inline flink-staging/flink-ml/src/test/scala/org/apache/flink/ml/evaluation/CrossValidationITSuite.scala]: It prints a line between the consecutive test runs, I just have it there so I can more easily see what is happening. The tests don't really do anything yet, just print results.

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala (hunk 2 above)
Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: sklearn returns an array of scores, as does spark. The mean on its own does not tell us much, it is very useful to see how much variance the scores have as well, i.e. if we have a robust model or not. Leaving it up to the user I think is a good choice here, so I would vote for returning a DataSet[Double], with all the scores (i.e. reduce->union on the scores array).

Comment (tillrohrmann) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: What else could one want to calculate from this? Maybe we can already calculate these two values for the user?

Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: Yes, that is the third option I have listed there, return mean and std. Me as a user I prefer to have all the scores.
What Spark does which I think is a good idea is calculate the mean and std and print it out, but give the user access to the actual scores as well.

Comment (tillrohrmann) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: I'm just wondering for what the user could use them? There is no way to establish an association with the specific fold for which it was calculated. Of course, you could calculate min and max values which might be interesting as well. Maybe this is justification enough.

Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: For me it's the simple act of just looking at them, you can get a pretty good idea of how the algorithm behaves, since it's usually just 5-10 values.

I'd like to provide more data rather than less to users, who can then make an informed decision.

Comment (thvasilo) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: sklearn defines the `score` function separately for the `Classifier` and `Regressor` mixins.

We can also try defining it in a trait `WithScore` but I'm not sure how that would work for chaining.

Comment (tillrohrmann) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: Ok, then we have to provide the means to easily calculate the mean and std from this data.

File: flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  override def folds[T](
+      input: DataSet[T],
+      seed: Long = new Random().nextLong()): Array[(DataSet[T], DataSet[T])] = {
+    val numFoldsF = numFolds.toFloat
+    (1 to numFolds).map { fold =>
+      val lb = (fold - 1) / numFoldsF
+      val ub = fold / numFoldsF
+      val validation = input.sampleBounded(lb, ub, complement = false, seed = seed)
+      val training = input.sampleBounded(lb, ub, complement = true, seed = seed)
+      (training, validation)
```
Comment (sachingoel0101) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: There is a potential problem here. I was doing the same thing to generate random splits of data based on an Array of relative proportions provided, but ideally, the union of these two sets should be the original data set. That is to say, validation and training should be mutually exclusive, and their union should be input.

Comment (tillrohrmann) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: Given that `input` is deterministic, this code will produce mutually exclusive validation and training data sets.

Comment (sachingoel0101) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: Ah yes. I see that now.

Comment (sachingoel0101) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: However, in case the parallelism of data is more than one, this can lead to problem. The random number sequence generated on every node would be the same, wouldn't it?
I printed all the random numbers generated and it looks like this: https://gist.github.com/sachingoel0101/ecde269af996fba7a39a

Further, for a parallelism of 2, the test itself fails.

Comment (tillrohrmann) [inline flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/evaluation/CrossValidation.scala]: The test shouldn't fail. Maybe there is an error then.

What one could do to have different sequences on each node is to xor the subtask id with the seed. But IMO this does not change the statistical properties of the sample because we don't know the underlying order of the elements. E.g. the underlying order of the element could be that way that we obtain the same sample set as with an identical seed and a different order.

**Comment by chobeat** [comment]
No news?

**Comment by zentol** [comment]
Closing since flink-ml is effectively frozen.
